### PR TITLE
gh-122749: Fix collections.Counter for non-standard counting

### PR DESCRIPTION
**Pull request by DirkHoffmann**
Avoid using 0 as default for empty counter bins.

The issue has been explained [here](https://github.com/python/cpython/issues/122749). 

The benefits can be demonstrated with this code (when comparing to the current implementation):

```python
from datetime import datetime
from collections import Counter
from time import sleep
import cmath

c = Counter()

ts0 = datetime.now()
sleep(.2)
ts1 = datetime.now()
dt = ts1-ts0

c.update({1:dt})
print(c)
c.update({1:dt})
print(c)
c.update({2:dt})
print(c)
c.update({2:dt, 3:dt-dt, 42:(dt+dt)/2})
print(c)

#Xtreme cases
sleep(.1)
ts1 = datetime.now()
dt = ts1-ts0
c.update({4:complex(1,1), 5: 18, 7:42.5, 42:dt})
c.update({4:complex(4,5)})
print(c)
```

* Issue: gh-122749

### CODE REVIEW

File: Lib/collections/__init__.py [hunk 1]
```diff
@@ -693,7 +693,11 @@ def update(self, iterable=None, /, **kwds):
                 if self:
                     self_get = self.get
                     for elem, count in iterable.items():
-                        self[elem] = count + self_get(elem, 0)
```
Comment (eendebakpt) [inline Lib/collections/__init__.py:696]: Since `self_get` is unused now, the definition two lines above this line can be removed.

**Comment by eendebakpt** [comment]
Some observations about the PR:

* The changes make the `Counter` behaviour more consistent, which is a good thing. 
* The changes are backwards incompatible. A minimal example with numpy:
```
import numpy as np # for numpy < 2.0
from collections import Counter
c=Counter()
c.update({'a': np.int16(2)})
c.update({'b': np.int16(2)})
print(c) # Counter({'a': 2, 'b': 2, 'c': 2})
print(type(c['a'])) # np.int16
print(type(c['b'])) # np.int32, will change to np.int16
```
I think the behavior change is acceptable, but would be good to mention in the news entry. 

* Performance impact seems to be small. The PR seems to be a tiny bit slower for `counter.update({'a': 1, 'b': 0})`, but the amount if close to the noise.

* The reason for the inconsistency is the fast path for an empty `Counter` added 16 years ago (https://github.com/python/cpython/commit/dd01f8f37b98af090098495200107dfaf92a3162). @rhettinger Is the fast path still needed? It seems to be only used for the very first element added to the counter and with this PR there is effectively a fast path when an element is not in the counter. So I suspect the gain is smaller, but the fast path still adds a bit of overhead for every call to `Counter.update`

**Comment by DirkHoffmann** [comment]
Thanks for these comparisons, @eendebakpt . I answer in two times, regarding compatibility and performance respectively. 

Regarding compatibility:

You omitted the a line like `c.update({c: 2})` for the update of the `'c'` counter in your example code, right?

I tested the "old" behaviour on python 3.10.11:

```
import numpy as np
from collections import Counter
c=Counter()
c.update({'a': np.int16(2)})
c.update({'b': np.int16(2)})
c # Counter({'a': 2, 'b': 2})
type(c['a']), type(c['b']) # (<class 'numpy.int16'>, <class 'numpy.int64'>)
```
(The same input comes out in two different formats.)

I draw the conclusions:
* The PR will store **np.int16** as **np.int16** and not change the type. I consider this an additional advantage.
* This is rather a flaw of **numpy**. When adding 0 to an **np.int16**, the result becomes **np.int64**, which is four times more memory.
* The result is apparently (according to your report) not well defined. (Which version of python were you using?) Once it became **np.int32**, on python 3.10.11 it becomes **np.int64**.
* Better store in the **Counter** bins the most exactly the type, which was given in **update()**. If there were usages that rely on an implicit numpy type change (which is not well defined, see above), they would be doomed anyway.
* (And such cases should be represented in the integration tests, shouldn't they?)

Illustration of numpy type changes:
```
>>> type(np.int16(2)), type(np.int16(2)+0), type(np.int16(2)+0), type(np.int32(2)+0)
(<class 'numpy.int16'>, <class 'numpy.int64'>, <class 'numpy.int64'>, <class 'numpy.int64'>)
```

But we are not discussing numpy here.

**Comment by DirkHoffmann** [comment]
Regarding performance:

I don't know how to install a package (numpy) for my development clone of python. But I considered that
```
self[elem] = count + self.get(elem, 0)
```
is equivalent to
```
if elem in self:
        self[elem] += count # or self[elem] = count + self[elem]
else:
        self[elem] = count + 0
```
also in terms of CPU time. The PR has even one addition less (`+ 0`). Are you sure it is slower? How did you test it?

I made a comparison and found the opposite result: The PR runs faster for me on the testcase
``` 
c = Counter()
for i in range(10000000):
    c.update({i: 42}) # filling an empty counter bin
    # c.update({i: 42}) # adding to an existing counter bin (in a second test)
```
on an i7-1165G7 CPU and obtained better (albeit not significantly, given the reproducibility) wall-clock time results (average and variance over 10 measurements:

|                                 | empty | empty+update 
| --- | --- | --- |
now (main branch) |  5.90±0.08 s|  11.84±.30 s|
PR (fix-issue-1222749) |  5.59±0.11 s|  11.61±.35 s|

But I agree with your conclusion that the differences are insignificant, given the uncertainty of the measurements.

**Comment by eendebakpt** [comment]
@DirkHoffmann The numpy behavior is not a bug or not well-defined. A description and rationale can be found it https://numpy.org/neps/nep-0050-scalar-promotion.html. The exact numpy behavior is not relevant for this PR though, the point I  wanted to make is that the PR changes the behavior of the `Counter.update`. 

The behavior change is a good one (it makes the `update` more consistent), but it is a change and therefore needs mentioning in the news entry and approval of a core developer.

File: Lib/collections/__init__.py [hunk 2]
```diff
@@ -691,9 +691,12 @@ def update(self, iterable=None, /, **kwds):
         if iterable is not None:
             if isinstance(iterable, _collections_abc.Mapping):
                 if self:
-                    self_get = self.get
                     for elem, count in iterable.items():
-                        self[elem] = count + self_get(elem, 0)
+                        if elem in self:
+                            self[elem] += count
```
Comment (eendebakpt) [inline Lib/collections/__init__.py:696]:
```suggestion
                            self[elem] = self[elem] + count
```
About the performance: results can depend on the system, platform and on the input type (e.g. adding a python `int`, a `bool` or `numpy.int16` to the counter). I tested with `%timeit counter.update({'a': 1})`. Both our tests show performance impact is small.

Perhaps you can even improve a bit by replacing `self[elem] += count` with `self[elem] = count + self[elem]`. A microbenchmark on my system:
```
import dis

def h():
    self={'a': 0}
    elem='a'
    for jj in range(1_000_000):
            self[elem] += 1

h()            

print(dis.dis(h, adaptive=True))

def g():
    self={'a': 0}
    elem='a'
    for jj in range(1_000_000):
            self[elem] = self[elem] + 1

g()            

print(dis.dis(g, adaptive=True))

%timeit h()
%timeit g()
```
Output of the timeit statements:
```
42.2 ms ± 831 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
36.6 ms ± 226 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Pull request closed without merging**